### PR TITLE
refactor Dockerfiles for PHP-FPM

### PR DESCRIPTION
**Pull request by henzigo**
#### Description, the reason for the PR

Compare PHP-FPM docker image:

| Type  | Compressed | Uncompressed
| ------------- | ------------- | ------------- |
| Monorepo - Debian | 616,98 MB  | 1.85 GB  |
| Monorepo - Alpine | 264,81 MB  | 790 MB  |
| Project Base - Debian  | 542.59 MB  | 1.63 GB  |
| Project Base - Alpine  | 174,22 MB  |  779 MB |

#### Fixes issues
... 

#### Check the appropriate checkboxes below, please

- [ ] [BC breaks](https://docs.shopsys.com/en/latest/contributing/backward-compatibility-promise/) 
- [ ] New feature 
- [x] I have read and signed [License Agreement for contributions](https://www.shopsys.com/license-agreement)

## Summary by CodeRabbit

- **New Features**
	- Enhanced Docker build processes for improved efficiency and clarity.
	- Introduced new development stages for better debugging and profiling options.
  
- **Bug Fixes**
	- Adjusted test cases for date-time formatting to meet specific spacing requirements.

- **Refactor**
	- Restructured Dockerfiles for clearer separation between production and development environments.
	- Streamlined CI/CD pipeline with improved job dependencies and logging capabilities.

- **Chores**
	- Removed obsolete scripts and commands to optimize the Docker image size.

----
:globe_with_meridians: Live Preview:
  - https://jg-dockerfile.odin.shopsys.cloud
  - https://cz.jg-dockerfile.odin.shopsys.cloud

### CODE REVIEW

**Comment by henzigo** [comment]
@coderabbitai review

File: packages/php-image/Dockerfile [hunk 1]
```diff
@@ -1,120 +1,80 @@
...
-# install Composer
-COPY ./docker-install-composer /usr/local/bin/docker-install-composer
-RUN chmod +x /usr/local/bin/docker-install-composer && docker-install-composer
-
-# install necessary php extensions for running application
-RUN docker-php-ext-configure gd --with-freetype --with-jpeg && \
+ARG PHP_VERSION=8.3
+
+# Node and Composer are already installed in the node_builder and composer_builder stages
+FROM composer:2.7 AS composer_builder
```
Comment (grossmannmartin) [inline packages/php-image/Dockerfile]: what about version `2.8` or `latest`?

File: packages/php-image/Dockerfile [hunk 2]
```diff
@@ -1,120 +1,80 @@
-ARG DEBIAN_VERSION=bullseye
+ARG LINUX_VERSION=alpine
```
Comment (grossmannmartin) [inline packages/php-image/Dockerfile]: _nitpick_ it's not a version, but a distribution

File: packages/php-image/Dockerfile [hunk 3]
```diff
@@ -1,120 +1,80 @@
...
+FROM composer:2.7 AS composer_builder
+FROM node:${NODE_MAJOR}-${LINUX_VERSION} AS node_builder
+
+
+FROM php:${PHP_VERSION}-fpm-${LINUX_VERSION} AS base
+
+
+# Build dependencies needed for installation and cleanup
+RUN apk add --no-cache --virtual .build-deps \
+    autoconf \
```
Comment (grossmannmartin) [inline packages/php-image/Dockerfile:50]: I like the idea of commentary about why some packages are installed.

File: packages/php-image/Dockerfile [hunk 4]
```diff
@@ -1,120 +1,80 @@
...
-# install locales and switch to en_US.utf8 in order to enable UTF-8 support
-# see http://jaredmarkell.com/docker-and-locales/ from where was this code taken
-RUN sed -i -e 's/# en_US.UTF-8 UTF-8/en_US.UTF-8 UTF-8/' /etc/locale.gen && \
-    locale-gen
-ENV LANG en_US.UTF-8
-ENV LANGUAGE en_US:en
-ENV LC_ALL en_US.UTF-8
-
-# add bash completion for phing
-COPY ./phing-completion /etc/bash_completion.d/phing
```
Comment (grossmannmartin) [inline packages/php-image/Dockerfile:101]: phing completion is quite useful even in the production environment.

if we decide to not include it in in the end, the phing-completion file could be deleted from the php-image package

File: project-base/app/docker/php-fpm/Dockerfile [hunk 5]
```diff
@@ -9,50 +9,85 @@ COPY ${project_root}/docker/php-fpm/php-ini-overrides.ini /usr/local/etc/php/php
 
 ########################################################################################################################
 
-FROM base as development
-
+FROM base AS production_builder
```
Comment (grossmannmartin) [inline project-base/app/docker/php-fpm/Dockerfile:12]: can you please change `as` to `AS` on the first line in this file to get rid of the warning?

File: packages/php-image/Dockerfile [hunk 6]
```diff
@@ -1,120 +1,80 @@
...
-        htop \
-        libfreetype6-dev \
-        libicu-dev \
-        libjpeg-dev \
-        libpng-dev \
-        libpq-dev \
-        librabbitmq-dev \
-        libzip-dev \
-        locales \
-        mc \
```
Comment (grossmannmartin) [inline packages/php-image/Dockerfile:50]: we discussed it internally and would like to include nano, vim, phing completion, and maybe even mc. 
it adds only around 45 MB, but the benefit is much greater

File: project-base/app/docker/php-fpm/Dockerfile [hunk 7]
```diff
@@ -9,50 +9,85 @@ COPY ${project_root}/docker/php-fpm/php-ini-overrides.ini /usr/local/etc/php/php
...
+&& curl -A "Docker" -o /tmp/blackfire-probe.tar.gz -D - -L -s https://blackfire.io/api/v1/releases/probe/php/alpine/$architecture/$version \
+&& mkdir -p /tmp/blackfire \
+&& tar zxpf /tmp/blackfire-probe.tar.gz -C /tmp/blackfire \
+&& mv /tmp/blackfire/blackfire-*.so $(php -r "echo ini_get ('extension_dir');")/blackfire.so \
+&& printf "extension=blackfire.so\nblackfire.agent_socket=tcp://blackfire:8307\n" > $PHP_INI_DIR/conf.d/blackfire.ini \
+&& rm -rf /tmp/blackfire /tmp/blackfire-probe.tar.gz
+
+RUN mkdir -p /tmp/blackfire \
+    && architecture=$(uname -m) \
+    && curl -A "Docker" -L https://blackfire.io/api/v1/releases/cli/linux/$architecture | tar zxp -C /tmp/blackfire \
```
Comment (grossmannmartin) [inline project-base/app/docker/php-fpm/Dockerfile:79]: should be here `linux` when above is `alpine`? 
also, previously it was installed in one step, is a reason to split it into two steps?

File: project-base/app/docker/php-fpm/Dockerfile [hunk 8]
```diff
@@ -9,50 +9,85 @@ COPY ${project_root}/docker/php-fpm/php-ini-overrides.ini /usr/local/etc/php/php
...
-RUN php phing build-deploy-part-1-db-independent clean
+RUN echo "source /etc/profile.d/bash_completion.sh" >> ~/.bash_profile && \
+    chown -R www-data:www-data /home/www-data /var/www/html
 
+USER www-data
+
+########################################################################################################################
+##################################### Development with blackfire ##########################
 ########################################################################################################################
+FROM development AS development_blackfire
```
Comment (grossmannmartin) [inline project-base/app/docker/php-fpm/Dockerfile:66]: what's the benefit of a separate stage for Blackfire? 
now, when we need Blackfire, we have to change the stage in docker-compose file. 

The documentation should be changed too (`blackfire-profiling.md`)

**Review by grossmannmartin — CHANGES_REQUESTED** [review]

File: packages/php-image/Dockerfile (hunk 1 above)
Comment (henzigo) [reply packages/php-image/Dockerfile]: fixed

File: packages/php-image/Dockerfile (hunk 2 above)
Comment (henzigo) [reply packages/php-image/Dockerfile]: fixed

File: project-base/app/docker/php-fpm/Dockerfile (hunk 5 above)
Comment (henzigo) [reply project-base/app/docker/php-fpm/Dockerfile:12]: fixed

File: packages/php-image/Dockerfile (hunk 4 above)
Comment (henzigo) [reply packages/php-image/Dockerfile:101]: I moved phing competion to the base image

File: project-base/app/docker/php-fpm/Dockerfile (hunk 7 above)
Comment (henzigo) [reply project-base/app/docker/php-fpm/Dockerfile:79]: It's from tutorial https://docs.blackfire.io/php/integrations/php-docker

File: packages/php-image/Dockerfile (hunk 6 above)
Comment (henzigo) [reply packages/php-image/Dockerfile:50]: Moved vim and nano to base image

File: packages/php-image/Dockerfile (hunk 3 above)
Comment (henzigo) [reply packages/php-image/Dockerfile:50]: fixed

File: project-base/app/docker/php-fpm/Dockerfile (hunk 8 above)
Comment (henzigo) [reply project-base/app/docker/php-fpm/Dockerfile:66]: I don't think that it's good idea to install things (even in development mode) if you don't use it. It's like install x-debug for everyone when 1 of 10 people uses it.

Comment (henzigo) [reply project-base/app/docker/php-fpm/Dockerfile:66]: Updated docs

**Review by grossmannmartin — APPROVED** [review]

**Pull request merged**